pystac/extensions/scientific.py

```diff
@@ -0,0 +1,272 @@
+
+
+class Publication:
+    """Helper for Publication entries."""
+    def __init__(self, doi: Union[str, Doi], citation: str) -> None:
+        if not isinstance(doi, Doi):
+            doi = Doi(doi)
+        self.doi = doi
+        self.citation = citation
+
+    def __eq__(self, other: _PublicationType) -> bool:
+        if not isinstance(other, Publication):
+            return NotImplemented
+
+        return self.doi == other.doi and self.citation == other.citation
+
+    def __repr__(self) -> str:
+        return f'<Publication doi={self.doi} target={self.citation}>'
+
+    def to_dict(self) -> Dict[str, str]:
+        return copy.deepcopy({'doi': self.doi.doi, 'citation': self.citation})
+
+    @staticmethod
+    def from_dict(d: Dict[str, str]) -> _PublicationType:
+        return Publication(d['doi'], d['citation'])
+
+    def get_link(self) -> link.Link:
+        url = self.doi.url()
+        return link.Link(CITE_AS, url)
+
+
+def remove_link(links: List[link.Link], pub: Publication):
+    url = pub.doi.url()
+    for i, a_link in enumerate(links):
+        if a_link.rel != CITE_AS:
+            continue
+        if a_link.target == url:
+            del links[i]
+            break
+
+
+class ScientificItemExt(base.ItemExtension):
+    """Add an citation and dois to a STAC Item."""
+    def __init__(self, an_item: item.Item) -> None:
+        self.item = an_item
+
+    def apply(self,
+              doi: Optional[Union[str, Doi]] = None,
+              citation: Optional[str] = None,
+              publications: Optional[List[Publication]] = None) -> None:
+        if doi:
+            self.doi = doi
+        if citation:
+            self.citation = citation
+        if publications:
+            self.publications = publications
+
+    @classmethod
+    def from_item(cls: _ScientificItemExtType, an_item: item.Item) -> _ScientificItemExtType:
+        return cls(an_item)
+
+    @classmethod
+    def _object_links(cls: _ScientificItemExtType) -> List:
+        return []
+
+    @property
+    def doi(self) -> Doi:
+        return self.item.properties.get(DOI)
+
+    @doi.setter
+    def doi(self, v: Union[str, Doi]) -> None:
+        if isinstance(v, Doi):
+            v = v.doi
+        self.item.properties[DOI] = v
+        url = DOI_URL_BASE + self.doi
+        self.item.add_link(pystac.link.Link(CITE_AS, url))
+
+    @property
+    def citation(self) -> str:
+        return self.item.properties.get(CITATION)
+
+    @citation.setter
+    def citation(self, v: str) -> None:
+        self.item.properties[CITATION] = v
+
+    @property
+    def publications(self) -> List[Publication]:
+        return [Publication.from_dict(pub) for pub in self.item.properties.get(PUBLICATIONS, [])]
+
+    @publications.setter
+    def publications(self, v: List[Publication]) -> None:
+        self.item.properties[PUBLICATIONS] = [pub.to_dict() for pub in v]
+        for pub in v:
+            self.item.add_link(pub.get_link())
```

Comment:
I think I may have been mistaken if I commented on publication links - the spec does not say that publications should be in the `links` of an item, only the item DOI link. This is added above in the doi setter. With this logic we have links that are marked as 'cite_as' for the doi of the data, and also any publications that reference it. I think the correct approach is to not have publications modify links, but have `doi` modify the single cite_as link.

Reply:
Rereading your comments above and this one, I will start cleaning up what I have.  I will remove the Doi class.  It's just a convenience wrapper that can easily be a function added later.

Reply:
Rereading the spec, I see this near the bottom defining a Publication:

> doi | string | DOI of a publication referencing the data.

And then after that is:

> For all DOI names respective DOI links SHOULD be added to the links section (see chapter "Relation types").

To me this implies that the publication DOI should be a link.  If it had said `sci:doi` should be added to the links, then I would have agreed that the pubs should not be in the links.

Does that make sense?

Reply:
Ah, good call - it seems like the publication DOIs should end up as links. And based on the language for the relation type, the `cite-as` rel type seems like the right way to go. This is confusing to me - to have several links with the rel type `cite-as` that point to either DOIs for the data DOIs for publications that point to the data - but that is how the spec reads now that you've clarified it for me. Thanks!

Reply:
`cite-as` does seem to have too many uses.  I guess we need to assume that the DOIs will be different and the code should search for cite-as links with the quoted doi string in the link.  Created https://github.com/radiantearth/stac-spec/issues/915 to discuss this.

Reply:
https://github.com/radiantearth/stac-spec/issues/915#issuecomment-722322144 said:

> The links were just intended as simplification for clients and users, mostly for the ones not supporting the scientific extension so that user could still follow links. It was just meant as a list of related dois to follow, not to transport any additional information or so.
> If two links would be the same, there should be just one link to it. The thing is that the DOI links can easily be generated if the client understands the sci extentsion/dois anyway so those clients don't really need the links.